pytential/symbolic/pde/scalar.py

```diff
@@ -252,108 +335,139 @@ def map_potentials(x):  # pylint:disable=function-redefined
 
         return (
                 map_potentials(
-                    sym.S(self.kernel, inv_sqrt_w_u, **kwargs))
-                - self.alpha
-                * map_potentials(
-                    sym.D(self.kernel,
-                        sym.S(self.laplace_kernel, inv_sqrt_w_u),
-                        **kwargs)))
-
-    def operator(self, u):
-        from sumpy.kernel import HelmholtzKernel, LaplaceKernel
-
+                    sym.S(self.kernel, inv_sqrt_w_u, **kwargs)
+                    )
+                - self.alpha * map_potentials(
+                    sym.D(self.kernel, laplace_s_inv_sqrt_w_u, **kwargs)
+                    )
+                )
+
+    def operator(self, u, **kwargs):
+        """
+        :param u: symbolic variable for the density.
+        :param kwargs: additional keyword arguments passed on to the layer
+            potential constructor.
+        """
         sqrt_w = self.get_sqrt_weight()
-        inv_sqrt_w_u = cse(u/sqrt_w)
-
-        knl = self.kernel
-        lknl = self.laplace_kernel
+        inv_sqrt_w_u = sym.cse(u/sqrt_w)
+        laplace_s_inv_sqrt_w_u = sym.cse(
+                sym.S(self.laplace_kernel, inv_sqrt_w_u, qbx_forced_limit=+1)
+                )
 
-        knl_kwargs = {}
-        knl_kwargs["kernel_arguments"] = self.kernel_arguments
-
-        DpS0u = sym.Dp(knl,  # noqa
-                cse(sym.S(lknl, inv_sqrt_w_u)),
-                **knl_kwargs)
+        kwargs["kernel_arguments"] = self.kernel_arguments
 
         if self.use_improved_operator:
-            Dp0S0u = -0.25*u + sym.Sp(  # noqa
-                    lknl,  # noqa
-                    sym.Sp(lknl, inv_sqrt_w_u, qbx_forced_limit="avg"),
-                    qbx_forced_limit="avg")
+            def Sp(density):
+                return sym.Sp(self.laplace_kernel,
+                        density,
+                        qbx_forced_limit="avg")
+
+            Dp0S0u = -0.25 * u + Sp(Sp(inv_sqrt_w_u))
 
+            from sumpy.kernel import HelmholtzKernel, LaplaceKernel
             if isinstance(self.kernel, HelmholtzKernel):
```

Comment:
Do you happen to have any reference for this? Is it in that Hackbusch book?

Reply:
It's just "right-precondition with a single layer and use a Calderón identity", see e.g. [here](https://andreask.cs.illinois.edu/cs598apk-f19/notes.pdf#page=278). The point is that the "improved" `Dp0S0u` is second-kind. But yes, I agree that this deserves a comment. :)

Reply:
Added some notes in 26d3c7f.